packages/metadata-common/src/MetadataEditor.tsx

```diff
@@ -100,6 +102,32 @@ export class MetadataEditor extends ReactWidget {
         if (!this.name) {
           this.title.label = `New ${this.schemaDisplayName}`;
         }
+        this.sortedFields = [];
+        const categories: string[] = [];
+        // Find categories of all schema properties
+        for (const schemaProperty in this.schema) {
```

Comment:
you should be able to simplify this `for` loop and remove the multiple `if` statements, like so:

```javascript
  const categories = new Set();

  for (const schemaProperty in this.schema) {
    categories.add(
        (this.schema[schemaProperty].uihints && this.schema[schemaProperty].uihints.category) || 
        null
    );
  }
```

and you can replace the `null` with whatever you want to represent no category (e.g., `''`, etc.)

Reply:
looking at this more closely. as a possible alternative, you could simplify further and actually replace the back-to-back `for` (line 108, line 120) with something like so:

```javascript
  const categories = { '_noCategory': [] }
  
  for (const schemaProperty in this.schema) {
    const cat = this.schema[schemaProperty].uihints && this.schema[schemaProperty].uihints.category;
    if (!cat) {
      categories['_noCategory'].push(schemaProperty)
    } else if (categories.hasOwnProperty(cat)) {
      categories[cat].push(schemaProperty)
    } else {
      categories[cat] = [schemaProperty]
    }
  }
```

this results in a `categories` that looks like so:
```json
{
  "_noCategory": [
    "description"
  ],
  "Kubeflow Pipelines": [
    "api_endpoint",
    "api_username",
    "api_password"
  ],
  "Cloud Object Storage": [
    "cos_endpoint",
    "cos_username",
    "cos_password",
    "cos_bucket"
  ]
}
```
which i believe gives you all the info to layout the fields in order by category...

Reply:
That looks great! Thanks!